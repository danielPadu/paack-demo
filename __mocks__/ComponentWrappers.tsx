import {NavigationContext} from '@react-navigation/native';
import React from 'react';
import {Provider} from 'react-redux/src';
import {PersistGate} from 'redux-persist/integration/react';
import {persistor} from '../appFiles/appStore/store';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import {defaultReducer} from '../appFiles/appStore/reducers/default';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
export const mockedStore = mockStore(defaultReducer);

export const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
  navigate: Function.prototype,
  setParams: Function.prototype,
  dispatch: Function.prototype,
  getParam: (param: {deliveryId?: string} & any, defaultValue: any) => {
    return defaultValue;
  },
};

export const ComponentWrappers = (props: {
  children: React.ReactNode;
  redux_store: any;
}) => {
  const {children, redux_store} = props;

  return (
    <>
      <Provider store={redux_store ?? mockedStore}>
        <PersistGate loading={null} persistor={persistor}>
          <NavigationContext.Provider value={navContext as any}>
            {children}
          </NavigationContext.Provider>
        </PersistGate>
      </Provider>
    </>
  );
};
