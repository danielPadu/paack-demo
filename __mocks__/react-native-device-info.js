export default {
  getModel: jest.fn(),
  getVersion: jest.fn(),
  getBuildNumber: jest.fn(),
  getBundleId: jest.fn(),
  hasNotch: jest.fn(),
};
