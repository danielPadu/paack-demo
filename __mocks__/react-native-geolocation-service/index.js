export default {
  addListener: jest.fn(),
  getCurrentPosition: () => ({
    coords: {
      accuracy: 0,
      altitude: 0,
      altitudeAccuracy: 0,
      heading: 0,
      latitude: 0,
      longitude: 0,
      speed: 0,
    },
  }),
  removeListeners: jest.fn(),
  requestAuthorization: jest.fn(),
  setConfiguration: jest.fn(),
  startObserving: jest.fn(),
  stopObserving: jest.fn(),
};
