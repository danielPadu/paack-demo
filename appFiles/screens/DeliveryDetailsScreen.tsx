import React, {useCallback, useEffect, useState} from 'react';
import {
  ActivityIndicator,
  SafeAreaView,
  ScrollView,
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  View,
} from 'react-native';
import {Button, Card} from 'react-native-ui-lib';
import {useDispatch, useSelector} from 'react-redux';
import {first} from 'rxjs/operators';
import {setAppActiveDeliveryID} from '../appStore/actions/appActions';
import useLocation from '../hooks/useLocation';
import type {DeliveryDetailsScreenProps} from '../navigation/RootNavigation';
import {
  getDeliveryDetails,
  updateDeliveryState,
} from '../services/rest/deliveryService';
import {LocationCard} from '../UI/components';
import {activityIndicator} from '../UI/components/buttonsActivityIndicator';
import {RenderField, RenderFieldTypes} from '../UI/components/RenderField';
import {DeliveryListItemTypes} from './types';

const DeliveryDetailsScreen = ({route}: DeliveryDetailsScreenProps) => {
  const {deliveryId} = route.params;
  const {active_delivery_id} = useSelector((state: any) => state.appReducer);
  const {location} = useLocation();
  const dispatch = useDispatch();
  const {
    wrapperStyle,
    containerStyle,
    pageTitleContainer,
    pageTitleStyle,
    showLocationButtonStyle,
    activeDeliveryText,
    cardContainer,
    clientContainer,
    makeActiveButtonStyle,
    makeActiveButtonLoadingStyle,
    clientText,
    cardStyle,
    emptyListText,
    changeStatusContainer,
    changeStatusButtonStyle,
    changeStatusButtonLabelStyle,
  } = styles;
  const [deliveryDetails, setDeliveryDetails] =
    useState<DeliveryListItemTypes | null>(null);
  const [customerInfo, setCustomerInfo] = useState<
    {title?: string; value?: string}[]
  >([]);
  const [apiRequest, setApiRequest] = useState(false);
  const settingApiReq = (val: boolean) => setApiRequest(val);
  const [showLocation, setShowLocation] = useState(false);
  const [makeActiveLoading, setMakeActiveLoading] = useState(false);
  useEffect(() => {
    if (deliveryId) {
      settingApiReq(true);
      getDeliveryDetails(deliveryId)
        .pipe(first())
        .subscribe({
          next: (data: any) => {
            setDeliveryDetails(data);
            const arrayData: {title: string; value: string}[] = [];
            Object.entries(data.customer).forEach(([key]) => {
              arrayData.push({
                title: `${key?.charAt(0)?.toUpperCase() + key?.slice(1)} : `,
                value: data.customer[key],
              });
            });
            setCustomerInfo(arrayData);
            settingApiReq(false);
          },
          error: () => {
            settingApiReq(false);
          },
        });
    }
  }, [deliveryId]);
  const getStatusColor = (value?: 'idle' | 'delivered' | 'undelivered') => ({
    color:
      value === 'delivered' ? 'green' : value === 'idle' ? 'orange' : 'red',
    fontSize: 12,
    fontWeight: '900',
  });

  const makeActiveLoadingIndicator = useCallback(() => {
    return makeActiveLoading ? (
      <ActivityIndicator
        style={makeActiveButtonLoadingStyle}
        size="small"
        color="#0000ff"
      />
    ) : null;
  }, [makeActiveButtonLoadingStyle, makeActiveLoading]);

  const onMakeActivePress = useCallback(() => {
    setMakeActiveLoading(true);
    if (deliveryDetails?.id) {
      dispatch(setAppActiveDeliveryID(deliveryDetails.id));
    }
    setMakeActiveLoading(false);
  }, [dispatch, deliveryDetails?.id]);

  const onChangeStatusPress = useCallback(
    (status: 'delivered' | 'undelivered') => {
      if (deliveryDetails?.id) {
        settingApiReq(true);
        updateDeliveryState(deliveryDetails.id, {
          delivery: {
            status: status,
            latitude: location?.latitude,
            longitude: location?.longitude,
          },
        })
          .pipe(first())
          .subscribe({
            next: (data: any) => {
              setDeliveryDetails(data);
              dispatch(setAppActiveDeliveryID(''));
              settingApiReq(false);
            },
            error: () => {
              settingApiReq(false);
            },
          });
      }
    },
    [deliveryDetails?.id, dispatch, location],
  );
  const noDeliverYDetails = useCallback(() => {
    return !apiRequest ? (
      <Text style={emptyListText}>
        {'No delivery details received at the moment'}
      </Text>
    ) : null;
  }, [apiRequest, emptyListText]);
  return (
    <SafeAreaView style={wrapperStyle}>
      <View style={containerStyle}>
        <View style={pageTitleContainer}>
          <Text style={pageTitleStyle}>{'Delivery Details Screen'}</Text>
        </View>
        <Button
          testID={'showLocationButton'}
          label={`${showLocation ? 'Hide' : 'Show'} device location`}
          onPress={() => setShowLocation(c => !c)}
          style={showLocationButtonStyle}
          bg-secondaryColor
          square
        />
        <ScrollView
          showsVerticalScrollIndicator={false}
          nestedScrollEnabled={false}>
          <View style={cardContainer}>
            {activityIndicator(apiRequest)}
            {deliveryDetails ? (
              <Card
                row={false}
                style={cardStyle}
                enableShadow={true}
                useNative
                activeOpacity={1}>
                <View style={clientContainer}>
                  <Text
                    style={clientText}>{`${deliveryDetails?.client} `}</Text>
                  {deliveryDetails?.id !== active_delivery_id &&
                    deliveryDetails?.delivery?.status !== 'delivered' && (
                      <Button
                        label="Make active"
                        disabled={active_delivery_id?.length > 0}
                        onPress={onMakeActivePress}
                        style={makeActiveButtonStyle}
                        iconOnRight
                        iconSource={makeActiveLoadingIndicator}
                        bg-primaryColor
                        square
                      />
                    )}
                  {deliveryDetails?.id === active_delivery_id && (
                    <Text style={activeDeliveryText}>{'Active delivery'}</Text>
                  )}
                </View>
                {customerInfo?.map((i, idx) => {
                  return <RenderField key={idx} {...(i as RenderFieldTypes)} />;
                })}
                <RenderField
                  title={'Status: '}
                  value={`${deliveryDetails?.delivery?.status} `}
                  valueExtraStyle={
                    getStatusColor(
                      deliveryDetails?.delivery?.status,
                    ) as StyleProp<TextStyle>
                  }
                />
                {deliveryDetails?.id === active_delivery_id && (
                  <View style={changeStatusContainer}>
                    <Button
                      label="Mark as delivered"
                      disabled={
                        deliveryDetails?.delivery?.status === 'delivered'
                      }
                      onPress={() => onChangeStatusPress('delivered')}
                      style={changeStatusButtonStyle}
                      iconOnRight
                      iconSource={makeActiveLoadingIndicator}
                      bg-successColor
                      square
                    />
                    <Button
                      label="Make as undelivered"
                      labelStyle={changeStatusButtonLabelStyle}
                      disabled={
                        deliveryDetails?.delivery?.status === 'delivered'
                      }
                      onPress={() => onChangeStatusPress('undelivered')}
                      style={changeStatusButtonStyle}
                      iconOnRight
                      iconSource={makeActiveLoadingIndicator}
                      bg-errorColor
                      square
                    />
                  </View>
                )}
              </Card>
            ) : (
              noDeliverYDetails()
            )}
          </View>
          {showLocation && <LocationCard />}
        </ScrollView>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  wrapperStyle: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    margin: 0,
    backgroundColor: 'white',
    color: 'black',
  },

  containerStyle: {
    flex: 1,
    width: '100%',
    justifyContent: 'space-between',
    paddingTop: isIOS ? 10 : 5,
  },
  pageTitleContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  pageTitleStyle: {
    textAlign: 'center',
    fontSize: 20,
    fontWeight: '900',
    color: 'navy',
    paddingVertical: 0,
    marginVertical: 0,
  },
  showLocationButtonStyle: {
    height: 40,
    width: 200,

    marginVertical: 10,
    alignSelf: 'center',
  },
  cardContainer: {
    margin: 0,
    marginTop: 2,
    padding: 0,
    marginHorizontal: 15,
    justifyContent: 'center',
    alignItems: 'center',
    paddingBottom: 30,
  },
  clientContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  },
  clientText: {
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
    paddingVertical: 10,
    flex: 1,
    textAlign: 'center',
    color: 'black',
    fontWeight: '600',
  },
  makeActiveButtonStyle: {
    height: 30,
    paddingVertical: 0,
    justifyContent: 'center',
    alignItems: 'center',
    marginLeft: 10,
  },
  makeActiveButtonLoadingStyle: {marginLeft: 10},
  activeDeliveryText: {
    textAlign: 'center',
    fontSize: 14,
    fontWeight: '900',
    color: 'green',
  },
  cardStyle: {
    flex: 1,
    marginTop: 5,
    padding: 10,
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'center',
    width: '100%',
    minHeight: 300,
    overflow: 'hidden',
    borderRadius: 20,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 6,
    },
    shadowOpacity: 0.39,
    shadowRadius: 8.3,

    elevation: 13,
  },

  emptyListText: {
    textAlign: 'center',
    fontSize: 20,
    fontWeight: '900',
    color: 'black',
  },
  changeStatusContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  },
  changeStatusButtonStyle: {
    height: 50,
    paddingVertical: 0,
    justifyContent: 'center',
    alignItems: 'center',
    maxWidth: '50%',
  },
  changeStatusButtonLabelStyle: {
    fontSize: 10,
    fontWeight: '900',
  },
});
export default DeliveryDetailsScreen;
