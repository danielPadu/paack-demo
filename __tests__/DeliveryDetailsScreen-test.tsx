import React from 'react';
import {fireEvent, waitFor} from 'react-native-testing-library';
import renderer, {
  act,
  ReactTestRenderer,
  ReactTestRendererJSON,
} from 'react-test-renderer';
import DeliveryDetailsScreen from '../appFiles/screens/DeliveriesScreen';
import {ComponentWrappers, mockedStore} from '../__mocks__/ComponentWrappers';
let store;
let component:
  | ReactTestRenderer
  | ReactTestRenderer[]
  | ReactTestRendererJSON
  | ReactTestRendererJSON[]
  | null;
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: () => ({app_allow_geolocation: true, last_location: {}}),
  useDispatch: () => jest.fn(),
}));
describe('checking DeliveryDetailsScreen component', () => {
  test('DeliveryDetailsScreen renders correctly', async () => {
    await act(async () => {
      await beforeEach(() => {
        store = mockedStore;

        component = renderer
          .create(
            <ComponentWrappers redux_store={store}>
              <DeliveryDetailsScreen />
            </ComponentWrappers>,
          )
          .toJSON();
        expect(component).toMatchSnapshot();
      });
    });
  });

  test('DeliveryDetailsScreen inner functionallity tests', async () => {
    await act(async () => {
      await beforeEach(async () => {
        store = mockedStore;

        component = renderer.create(
          <ComponentWrappers redux_store={store}>
            <DeliveryDetailsScreen />
          </ComponentWrappers>,
        );
        const {getByTestId} = component;
        const mockedHandler = jest.fn();
        const button = getByTestId('showLocationButton');
        button.props.onPress = mockedHandler;
        fireEvent.press(button);
        await waitFor(() => expect(mockedHandler).toHaveBeenCalled());
        await waitFor(() => expect(component).toHaveProperty('wrapperStyle'));
      });
      await afterEach(() => {
        // cleanup on exiting
        component?.remove?.();
        component = null;
      });
    });
  });
});
// test('renders correctly', () => {
//   const tree = renderer
//     .create(
//       <DeliveryDetailsScreen
//         navigation={
//           navigationRef.current as unknown as NativeStackNavigationProp<
//             RootStackParamList,
//             'DeliveriesScreen'
//           >
//         }
//         route={
//           navigationRef.current as unknown as RouteProp<
//             RootStackParamList,
//             'DeliveriesScreen'
//           >
//         }
//       />,
//     )
//     .toJSON();
//   expect(tree).toMatchSnapshot();
// });
