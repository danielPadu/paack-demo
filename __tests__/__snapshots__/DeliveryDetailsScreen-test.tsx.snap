// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checking DeliveryDetailsScreen component DeliveryDetailsScreen inner functionallity tests 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "color": "black",
      "flex": 1,
      "justifyContent": "space-between",
      "margin": 0,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "paddingTop": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "navy",
            "fontSize": 20,
            "fontWeight": "900",
            "textAlign": "center",
          }
        }
      >
        Deliveries List
      </Text>
    </View>
    <RCTScrollView
      nestedScrollEnabled={false}
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "justifyContent": "center",
              "margin": 0,
              "marginHorizontal": 15,
              "marginTop": 2,
              "padding": 0,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "black",
                "fontSize": 20,
                "fontWeight": "900",
                "textAlign": "center",
              }
            }
          >
            No deliveries at the moment
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
